bpo-30681: Support invalid date format or value in email Date header

sim0nx opened the pull request:
I am re-submitting an older PR which was abandoned but is still relevant, #10783 by @timb07.

The issue being solved (https://bugs.python.org/issue30681) is still relevant. The original PR #10783 was closed as
the final request changes were not applied and since abandoned.

In this new PR I have re-used the original patch plus applied both comments from the review, by @maxking and @pganssle.


For reference, here is the original PR description:
In email.utils.parsedate_to_datetime(), a failure to parse the date, or invalid date components (such as hour outside 0..23) raises an exception. Document this behaviour, and add tests to test_email/test_utils.py to confirm this behaviour.

In email.headerregistry.DateHeader.parse(), check when parsedate_to_datetime() raises an exception and add a new defect InvalidDateDefect; preserve the invalid value as the string value of the header, but set the datetime attribute to None.

Add tests to test_email/test_headerregistry.py to confirm this behaviour; also added test to test_email/test_inversion.py to confirm emails with such defective date headers round trip successfully.

This pull request incorporates feedback gratefully received from @bitdancer, @brettcannon, @Mariatta and @warsaw, and replaces the earlier PR #2254.

https://bugs.python.org/issue30681


<!-- issue-number: [bpo-30681](https://bugs.python.org/issue30681) -->
https://bugs.python.org/issue30681
<!-- /issue-number -->

Automerge-Triggered-By: GH:warsaw

sim0nx committed:
bpo-30681: Support invalid date format or value in email Date header

Re-submitting an older PR which was abandoned but still relevant.

warsaw commented on Doc/library/email.utils.rst:
```diff
@@ -124,7 +124,10 @@ of the new API.
 
    The inverse of :func:`format_datetime`.  Performs the same function as
-   :func:`parsedate`, but on success returns a :mod:`~datetime.datetime`.  If
+   :func:`parsedate`, but on success returns a :mod:`~datetime.datetime`;
```
I'm looking at this now during the Python core sprint.  Thanks for resurrecting this PR.

I find this description a bit confusing because it's saying that the function either returns `None` or raises `ValueError` when parsing an invalid date.  That makes it difficult to know when I'll get `None` and when I'll get an exception.  I think you're saying that only some bogus strings will raise `ValueError`, e.g. those where the hour is invalid.  I also don't see where the tests you're adding include the test for `None` return.

I'm not sure what to suggest, because I'm not sure what the intended semantics *should* be.

sim0nx replied:
I actually don't see how *parsedate_to_datetime* would ever return None. It is *_parsedate_tz* which returns **None** on bogus input, in which case *parsedate_to_datetime* raises a **TypeError**.
This is also covered in the tests, so those should be fine.

In order to continue I suggest to fix the documentation on *parsedate_to_datetime*, remove the mention of it returning None and replacing it with it possibly returning **TypeError** in case of an invalid date input.

warsaw replied:
As I mentioned in the ticket, I agree with you; `parsedate_to_datetime()` can't return None, so let's remove that from the documentation.  I also suggest checking for `None` return from `_parsedate_tz()` and turning that into an explicit raise of `ValueError` rather than relying on the implicit `TypeError` from a failing tuple unpack.  Then the docs should say that `ValueError` is raised on an invalid date input.

sim0nx replied:
Thanks for your comments. I fully agree with your points, that makes it definitely cleaner.
I implemented them in my latest commit.

sim0nx committed:
Fix documentation for function *parsedate_to_datetime*; mention that it raises a TypeError on invalid input instead of it returning None

sim0nx committed:
Implement comments from code review

- _parsedate_tz should explicitely return None
- parsedate_to_datetime should check for None as return value by _parsedate_tz
   and in that case raise ValueError
- Fix documentation and tests accordingly
